LLM-controlled animation duration logic added

hasantahabagci opened the pull request:
### Summary

Implemented support for the LLM to decide animation duration for more natural pacing.

---

### 🔧 Changes

* [x] **Server (animation\_handler.py)**

  * Added optional `duration` param (0.5–10s, default 3s)
  * Validates input + sends duration in event payload

* [x] **Client (TalkingHeadWrapper.tsx)**

  * Handles duration from server
  * Applies duration to gestures/moods
  * Falls back to defaults if missing

fabiocat93 commented on src/client/react/src/components/avatarInteraction/TalkingHeadWrapper.tsx:
```diff
@@ -133,12 +133,15 @@ const TalkingHeadWrapper = forwardRef<object, Props>((props, ref) => {
       if (typeof val === "string") {
-        head.playGesture(val);
+        // Check if it's a file path
+        if (val.startsWith("/animations/")) {
```
@hasantahabagci agree

fabiocat93 commented on src/server/authorization/authorized_users.json:
this file shouldn't actually be public. my bad! do you mind removing it? thank you!

hasantahabagci replied:
Of course 👍

fabiocat93 commented on src/server/bot/components/llm_tools/animation_handler.py:
```diff
@@ -127,12 +127,18 @@ def build_animation_tools_schema(allowed_animations: List[str]) -> "FunctionSche
         return FunctionSchema(
             name="trigger_animation",
-            description="Trigger an avatar animation (only one at a time).",
+            description="Trigger an avatar animation (only one at a time). The LLM determines the duration for gestures and moods based on the flow of the conversation. If duration is not specified, select an appropriate value according to context.",
```
you are speaking to the LLM here. Please phrase the request accordingly

fabiocat93 commented on src/server/bot/components/llm_tools/animation_handler.py:
```diff
@@ -127,12 +127,18 @@ def build_animation_tools_schema(allowed_animations: List[str]) -> "FunctionSche
+                    "type": "number",
+                    "minimum": 0.5,
+                    "maximum": 10.0,
+                    "description": "Duration in seconds for the animation. Default is 3 seconds. Use shorter durations (0.5-3s) for quick gestures, longer durations (4-10s) for sustained moods or complex animations.",
```
wouldn't be better to have default durations animation-specific?

hasantahabagci replied:
I think so too, but I wanted to leave it up to LLM to decide. I didn't do anything specific to the animation so he could decide how long to animate "i_dont_know" or "hands_up." It might be better if I also added a prompt like "decide on animation." What do you think about this?

fabiocat93 reviewed (commented):
thank you @hasantahabagci ! I have left some comments here. Can you pls 1) address them and 2) attach a video demo of some animations so that we can vote how good they look?

hasantahabagci commented:
> thank you @hasantahabagci ! I have left some comments here. Can you pls 1) address them and 2) attach a video demo of some animations so that we can vote how good they look?

Sure, I'm addressing them and preparing video demo. Thanks for your comments.

hasantahabagci commented:
> > thank you @hasantahabagci ! I have left some comments here. Can you pls 1) address them and 2) attach a video demo of some animations so that we can vote how good they look?
> 
> Sure, I'm addressing them and preparing video demo. Thanks for your comments.

Here is the some animations demo: https://drive.google.com/file/d/1YfajrerVHYbusZ7wFZw3GAE2m2NWx4st/view?usp=sharing

hasantahabagci commented:
> > > > thank you @hasantahabagci ! I have left some comments here. Can you pls 1) address them and 2) attach a video demo of some animations so that we can vote how good they look?
> 
> > > 
> 
> > > 
> 
> > > Sure, I'm addressing them and preparing video demo. Thanks for your comments.
> 
> > 
> 
> > Here is the some animations demo: https://drive.google.com/file/d/1YfajrerVHYbusZ7wFZw3GAE2m2NWx4st/view?usp=sharing
> 
> 
> 
> Thank you @hasantahabagci ! I appreciate the concept, but have some concerns about the implementation. For instance, 
> 
> - I am not sure it actually follows the duration instructions. In the first exchange, it was supposed to dance for 4 seconds, but it lasted around 10 seconds. Is the duration parameter actually used?

Parameter is actually used but I could not understand that (maybe caused by some latency issues). I tested it for 30 seconds or different values it's working. maybe there is a problem with fbx animations, because I did not this like problem in other animations. Let me check for it.

> 
> - Why didn't it play the dancing animation for 1 minute?

Actually it caused by maximum animation duration parameter. I have set it to 10 seconds so that it doesn't do more. Just I was testing I give 30 seconds as parameter and it did it

> 
> 
> 
> cceing @batwood-1 since I am curious to hear his thoughts

hasantahabagci commented:
> > Parameter is actually used but I could not understand that (maybe caused by some latency issues). I tested it for 30 seconds or different values it's working. maybe there is a problem with fbx animations, because I did not see this like problem in other animations. Let me check for it.
> > >
> 
> does the issue maybe happen when the llm wants the animation to run for longer than the original duration? if that's the case, we should address this. or at least debug it
> 
> > Actually it caused by maximum animation duration parameter. I have set it to 10 seconds so that it doesn't do more. Just I was testing I give 30 seconds as parameter and it did it
> 
> same as above. this should be debugged

Just fixed it. I changed the animation playing function from `playPose()` to `playAnimation()`. Now working good. Also added thinking animation and tried on it.

fabiocat93 reviewed (approved):
@hasantahabagci good job!
